test/plugin/ckpttimer/ckpttimer.cpp

```diff
@@ -0,0 +1,243 @@
+  sa.sa_sigaction = timeout_handler;
+  sigemptyset(&sa.sa_mask);
+  if (sigaction(g_sig_num, &sa, NULL) == -1) {
+      handleError("sigaction");
+  }
+}
+
+static timer_t
+make_timer()
+{
+  timer_t timerid = 0;
+  struct sigevent sev;
+
+  setup_handler();
+
+  sev.sigev_notify = SIGEV_SIGNAL;
+  sev.sigev_signo = g_sig_num;
+  sev.sigev_value.sival_ptr = &timerid;
+  if (timer_create(CLOCKID, &sev, &timerid) == -1) {
+      handleError("timer_create");
+  }
+
+  return timerid;
+}
+
+static void
+start_stop_timer(timer_t timerid, long long interval, bool start)
+{
+  struct itimerspec its;
+
+  if (start) {
+    its.it_value.tv_sec = interval / 1000000000;
+    its.it_value.tv_nsec = interval % 1000000000;
+  } else {
+#ifdef CKPTTIMER_PLUGIN_DEBUG
+    if (timer_gettime(timerid, &its) == 0) {
+      DPRINTF("it_value(%lld, %lld), it_interval(%lld, %lld)\n", its.it_value.tv_sec, its.it_value.tv_nsec, its.it_interval.tv_sec, its.it_interval.tv_nsec);
+    } else {
+      handleError("timer_gettime");
+    }
+#endif
+    its.it_value.tv_sec = 0;
+    its.it_value.tv_nsec = 0;
+  }
+
+  its.it_interval.tv_sec = its.it_value.tv_sec;
+  its.it_interval.tv_nsec = its.it_value.tv_nsec;
+
+  if (timer_settime(timerid, 0, &its, NULL) == -1) {
+    handleError("timer_settime");
+  }
+#ifdef CKPTTIMER_PLUGIN_DEBUG
+  if (timer_gettime(timerid, &its) == 0) {
+    DPRINTF("After disabling: it_value(%lld, %lld), it_interval(%lld, %lld)\n", its.it_value.tv_sec, its.it_value.tv_nsec, its.it_interval.tv_sec, its.it_interval.tv_nsec);
+  } else {
+    handleError("timer_gettime");
+  }
+#endif
+}
+
+extern "C" void
+dmtcp_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
+{
+  static timer_t timerid = 0;
+  static int doneInitialization = 0;
+  sigset_t mask;
+
+  switch (event) {
+    case DMTCP_EVENT_INIT:
+      {
+        if (!doneInitialization) {
+          get_and_save_envvars();
+
+          /* Create a timer */
+          timerid = make_timer();
+          sigemptyset(&mask);
+          sigaddset(&mask, g_sig_num);
+          doneInitialization = 1;
+        }
+
+        DPRINTF("The plugin containing %s has been initialized.\n", __FILE__);
+        break;
+      }
+    case DMTCP_EVENT_WRITE_CKPT:
+      {
+        DPRINTF("*** The plugin is being called before checkpointing. ***\n");
+        /* Unblock the timer signal, and then start the timer */
+        if (sigprocmask(SIG_UNBLOCK, &mask, NULL) == -1)
+          handleError ("sigprocmask");
+        start_stop_timer(timerid, g_interval, START_TIMER);
+        break;
+      }
+    case DMTCP_EVENT_THREADS_SUSPEND:
+      {
+        DPRINTF("*** The plugin is being called after suspend. ***\n");
+        break;
+      }
+    case DMTCP_EVENT_RESUME:
+      {
+        DPRINTF("*** The plugin has now been checkpointed. ***\n");
+#ifdef CKPTTIMER_PLUGIN_TEST
+        static int dummy = 0;
+        while (!dummy);
+#endif
+        start_stop_timer(timerid, g_interval, STOP_TIMER);
```

Comment:
Why do we have to stop the timer twice? Both in THREADS_RESUME and RESUME? Wouldn't only one of them suffice?